sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +408,284 @@ class GpuSequenceMeta(
   }
 
   override def convertToGpu(): GpuExpression = {
-    if (expr.stepOpt.isDefined) {
-      val Seq(start, stop, step) = childExprs.map(_.convertToGpu())
-      GpuSequenceWithStep(start, stop, step, expr.timeZoneId)
-    } else {
-      val Seq(start, stop) = childExprs.map(_.convertToGpu())
-      GpuSequence(start, stop, expr.timeZoneId)
-    }
+    val (startExpr, stopExpr, stepOpt) = if (expr.stepOpt.isDefined) {
+        val Seq(start, stop, step) = childExprs.map(_.convertToGpu())
+        (start, stop, Some(step))
+      } else {
+        val Seq(start, stop) = childExprs.map(_.convertToGpu())
+        (start, stop, None)
+      }
+    GpuSequence(startExpr, stopExpr, stepOpt, expr.timeZoneId)
   }
 }
 
-object GpuSequenceUtil {
+object GpuSequenceUtil extends Arm {
 
-  def numberScalar(dt: DataType, value: Int): Scalar = dt match {
-    case ByteType => Scalar.fromByte(value.toByte)
-    case ShortType => Scalar.fromShort(value.toShort)
-    case IntegerType => Scalar.fromInt(value)
-    case LongType => Scalar.fromLong(value.toLong)
-    case _ =>
-      throw new IllegalArgumentException("wrong data type: " + dt)
+  def isAllTrue(col: ColumnVector): Boolean = {
+    assert(DType.BOOL8 == col.getType)
+    if (col.getRowCount == 0) {
+      return true
+    }
+    if (col.hasNulls) {
+      return false
+    }
+    withResource(col.all()) { allTrue =>
+      // Guaranteed there is at least one row and no nulls so result must be valid
+      allTrue.getBoolean
+    }
   }
-}
-
-/** GpuSequence without step */
-case class GpuSequence(start: Expression, stop: Expression, timeZoneId: Option[String] = None)
-    extends GpuBinaryExpression with TimeZoneAwareExpression {
-
-  override def left: Expression = start
-
-  override def right: Expression = stop
-
-  override def dataType: DataType = ArrayType(start.dataType, containsNull = false)
 
-  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
-    copy(timeZoneId = Some(timeZoneId))
+  /** (Copied most of the part from GpuConditionalExpression) */
+  def gather(predicate: ColumnVector, toBeGathered: ColumnVector): ColumnVector = {
+    // convert the predicate boolean column to numeric where 1 = true
+    // and 0 (or null) = false and then use `scan` with `sum` to convert to indices.
+    //
+    // For example, if the predicate evaluates to [F, null, T, F, T] then this
+    // gets translated first to [0, 0, 1, 0, 1] and then the scan operation
+    // will perform an exclusive sum on these values and produce [0, 0, 0, 1, 1].
+    // Combining this with the original predicate boolean array results in the two
+    // T values mapping to indices 0 and 1, respectively.
+    val boolAsInt = withResource(Scalar.fromInt(1)) { one =>
+      withResource(Scalar.fromInt(0)) { zero =>
+        predicate.ifElse(one, zero)
+      }
+    }
```

Comment:
This should be refactored as part of the new utilities class, same as `boolToInt` in `GpuConditionalExpression`.

Reply:
This is no longer needed